backend/app/utils/auth/AuthActionBuilder.scala

```diff
@@ -33,60 +34,75 @@ class DefaultAuthActionBuilder(val controllerComponents: ControllerComponents, f
+    val isVerificationExpired = token.verificationExpiry <= now
+    for {
+      maybeDbUser <- if (isVerificationExpired) getUser(token.user.username).asFuture else Future.successful(Left(AuthenticationFailure("Verification hasn't expired", reportAsFailure = true)))
+      result <- block(new AuthenticatedRequest(token.user, request))
+    } yield {
+      if (isVerificationExpired) {
+        maybeDbUser match {
+          case Right(user) if user.invalidationTime.exists(token.issuedAt < _) =>
+            // The user has logged out
+            val msg = s"Authenticated failed because token was issued before database invalidation time"
+            logger.warn(token.user.asLogMarker, msg)
+            Left(AuthenticationFailure(msg, reportAsFailure = true))
+          case Left(failure) =>
+            logger.error(token.user.asLogMarker, "Authentication failed because user was not found in DB", failure.toThrowable)
+            Left(failure)
+          case Right(_) =>
+            val verificationExpiry = now + maxVerificationAge.toMillis
+            val expiryDateTime = LocalDateTime.ofInstant(
+              Instant.ofEpochMilli(verificationExpiry),
+              ZoneId.systemDefault()
+            )
+            logger.info(token.user.asLogMarker, s"Authentication succeeded, verification expired but token renewed. New verification expiry: ${expiryDateTime}")
             Right(result
               .refreshJwtSession
+              .addingToJwtSession(Token.VERIFICATION_EXPIRY_KEY, verificationExpiry)
               .addingToJwtSession(Token.REFRESHED_AT_KEY, now)
             )
-          }
         }
-
-      case JsSuccess(token, _) => {
-        val msg = s"Token is older than $maxLoginAge"
-        logger.info(token.user.asLogMarker, msg)
-        Future.successful(Left(AuthenticationFailure(msg, reportAsFailure = false)))
+      } else {
+        logger.info(token.user.asLogMarker, s"Authentication succeeded")
+        Right(result
+          .refreshJwtSession
+          .addingToJwtSession(Token.REFRESHED_AT_KEY, now)
+        )
       }
+    }
+  }
 
-      case JsError(errors) => {
-        val msg = s"Failed to parse token: $errors"
-        logger.warn(msg)
-        Future.successful(Left(AuthenticationFailure(msg, reportAsFailure = true)))
+  private[auth] def invokeBlockWithTime[A](request: Request[A], block: (UserIdentityRequest[A]) => Future[Result],
+                                           now: Long): Future[Either[Failure, Result]] = {
+    implicit val implicitReq = request
+
+    // Why can't we do request.jwtSession.isEmpty?
+    // Because request.jwtSession gets initialised with some default properties if there's no
+    // actual JWT token in the request. This seems a little odd to me, but it's in the pdi.jwt
+    // library so we can't control this behaviour.
+    if (request.headers.get(JwtSession.REQUEST_HEADER_NAME).isEmpty) {
```

Comment:
FYI here's where it sets `request.jwtSession`:

https://github.com/jwt-scala/jwt-scala/blob/v4.3.0/play/src/main/scala/JwtPlayImplicits.scala#L19

Reply:
and here's where the `JwtSession` case class gets initialised with default stuff if called with no params: https://github.com/jwt-scala/jwt-scala/blob/v4.3.0/play/src/main/scala/JwtSession.scala#L150